pravega/pravega
# Issue 3417: (SegmentStore) Self Test additions for Table Segments

**Change log description**  
- Added the ability to performance test Table Segments in the SelfTest tool.
    - Updates & Conditional Updates
    - Removals & Conditional Removals
    - Get
    - Iterators

**Purpose of the change**  
Fixes #3417.

**What the code does**  
The `StoreAdapter` was extended with APIs for Tables (update, remove, get, iterate). As of this PR, only the SegmentStoreDirect implementation of that class has support for these. A future PR may add support via the client so we can test end-to-end.

The test accepts familiar inputs: number of operations, write size, number of producers and producer parallelism. It also introduces a few more: update/remove ratio, update/create ratio and consumer parallelism.

The test executes a number of operations as requested via the input. A test can execute either conditional updates/removes or unconditional updates/removes (the two types cannot be intermixed and have to be defined via the command-line input). At each operation it will decide whether to create a new key, update an existing key, remove an existing key or remove an inexistent key. This is decided by using a random number generator and the update/remove ratio and update/create ratio as defined in the inputs.

All updates are queued up for the Table Consumer to process. The Table Consumer runs in a loop where it picks all outstanding keys from this queue, dedupes them, groups them by target Table and issues batch-gets for those groups. The degree of parallelism is determined by the consumer parallelism input.

Throughout the test, it outputs ongoing stats such as number of operations (of each type) executed and "throughput". At the end, it outputs a familar looking table with latency stats by operation type.


**How to verify it**  
Not touching core code. You may check out the code and run the self test using the new args.
Sample args to start a self test run:
`-Dtarget=SegmentStoreTable -Dbkc=0 -Dc=1 -Ds=1 -Dsc=1 -Dws=100 -Do=100000 -Dp=1000 -Dselftest.tableConditionalUpdates=true`
